### ISSUE TITLE
Possibility to loose whole work while saving file (on tablet)

### ISSUE THREAD

**Issue by dominikzalewski** [OpenOrienteering/mapper#841]
### Steps to reproduce
1. Prepare large enough source OO Mapper file (6MB was my case). It helps to start with XMAP, since the file is larger by design.
2. Open the file (tablet is where I discovered it).
3. Click save. While save is ongoing, unplug the battery, unplug the power from your computer, send kill -9 to the program, or close it from a System Manager. Basically do whatever it takes to make the app shutdown ungracefully (without any warning from the OS). Running out of battery while operating outdoors in under zero celsius is how I discovered the issue.

### Actual behaviour
Both the file on the disc and the GPX track file gets zeroed on the disk/SD card. There is no backup made by the program, so one looses every bit of work that was done.

From what I suspect, the code is written simply to open both files for writing (effectively zeroing them on the disk/SD card) and then dump whatever is in the memory to those files. However, this approach lacks transactionality - the ability to back off to the previous state, when something goes wrong during the save operation.

### Expected behaviour
OO Mapper should create a backup of a previous save, before starting a new save. This way the save operation will be transactional. If something happens during the save, the user will have the ability to restore to a previous save from the backup file(s) on the disk.

### Configuration
Mapper Version: 0.6.7 (but probably all version are affected)
Operating System: Android (but the issue is probably system independent)

### Workaround
Install mgit (or any other git client) on a tablet and use it to commit the file while operating outdoors. This effectively emulates backups as described in 'Expected behaviour'.

### My story
The sad fact is that I already had a VCS configured on my tablet and was using it to sync files between my tablet and computer. I even wanted to make a git backup of the files (commit), but something distracted me. Because my tablet was showing me 46% of battery power, I did not want to connect a power bank, so that I don't have dangling power wires. However, I have forgotten that I'm operating on a near freezing temperature, and this makes the batteries in mobile devices go south without any warning even when the usage of battery is not critical. I clicked save after finishing a map fragment that took me 3 hours to complete. I could not believe that I have lost all the work, when I relaunched my tablet with power bank connected.

I understand that this issue was really hard to stumble upon. Normally if someone gets a warning that the battery is low, he probably connects a charger/power bank. And it was also pretty unlikely that the user would be operating till 0% that would only cause the device to shutoff while saving.

**Comment by dg0yt** [comment]
Maps are already saved transactional: They are written using [QSaveFile](http://doc.qt.io/qt-5/qsavefile.html#details). The original file is not touched until writing is complete and we explicitliy call commit(). Cf.
https://github.com/OpenOrienteering/mapper/blob/master/src/core/map.cpp#L638

**Comment by dg0yt** [comment]
And for the record, the autosave file is not removed until saving the map was successful.

**Comment by dominikzalewski** [comment]
Perhaps on Android getDirectWriteFallback() returns true?

http://doc.qt.io/qt-5/qsavefile.html#setDirectWriteFallback

**Comment by dg0yt** [comment]
No, I do not disregard the issue. That's why I left it in open state for now, despite the "invalid" label. However, I believe that Mapper worked correctly, i.e. as described in "Expected behaviour".

The directWriteFallback feature is an exceptional behaviour which needs to be enabled explicitly by the developer: It is against the promise of class QSaveFile. But it is not touched on Android or by Mapper. Source code browser: https://code.woboq.org/qt5/qtbase/src/corelib/io/qsavefile.h.html#QSaveFile

Given the low temperatures, the energy supply may have dropped more rapidly that anticipated by the operating system. The out-of-battery situation may have triggered undefined behaviour in the hardware or operating system of your device. Mapper is probably being signaled write success by the operating system while the data (or meta data) is not yet physically saved to the device. This is how operating systems improve (perceived) performance.
